modules/core/src/action.ts

```diff
@@ -1,19 +1,45 @@
 import {curry2} from 'ts-curry'
+import {isAction} from './isAction'
 import {isNil} from './isNil'
 
 type AType = string | number
 
-/**
- * Interface for Action
- * @interface
- */
-export interface Action<V, T = AType> {
-  readonly type: T
-  readonly value: V
+type func<a, b> = (val: a, state: b) => b
+
+type sAction<V, T extends string | number> = Action<V, T>
+
+type foldSpec<V, S> = V extends sAction<any, infer K>
+  ? {
+      [k in K]: V extends Action<infer Val, k>
+        ? Val extends Action<unknown, string>
+          ? (foldSpec<Val, S> | func<Val, S>)
+          : (val: Val, state: S) => S
+        : never
+    }
+  : never
+
+export class Action<V, T = AType> {
+  private constructor(readonly value: V, readonly type: T) {}
+  static of<T extends AType, V>(value: V, type: T) {
```

Comment:
- Type parameters can be as they are, but parameters need to be `type` first and then `value`.